src/Relation/Nullary/Finite/Setoid.agda

```diff
@@ -0,0 +1,56 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Notions of finiteness for setoids
+------------------------------------------------------------------------
+
+{-# OPTIONS --safe --cubical-compatible #-}
+
+module Relation.Nullary.Finite.Setoid where
+
+open import Data.Fin.Base using (Fin)
+open import Data.Nat.Base using (ℕ)
+open import Function.Bundles using (Injection; Surjection; Inverse)
+open import Level using (Level; _⊔_) renaming (suc to lsuc)
+open import Relation.Binary.Bundles using (Setoid)
+open import Relation.Binary.PropositionalEquality.Properties using (setoid)
+
+private
+  variable
+    c ℓ c′ ℓ′ : Level
+
+record IsStronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ) where
+   field
+     inv : Inverse X (setoid (Fin n))
+
+record StronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ) where
+   field
+     isFinite : IsStronglyFinite X n
+
+record Subfinite (X : Setoid c ℓ) : Set (c ⊔ ℓ) where
+  field
+    size : ℕ
+    inj : Injection X (setoid (Fin size))
+
+record FinitelyEnumerable (X : Setoid c ℓ) : Set (c ⊔ ℓ) where
+  field
+    size : ℕ
+    srj : Surjection (setoid (Fin size)) X
+
+record InjectsIntoFinitelyEnumerable (X : Setoid c ℓ) c′ ℓ′
+       : Set (c ⊔ ℓ ⊔ lsuc (c′ ⊔ ℓ′)) where
+  field
+    Apex : Setoid c′ ℓ′
+    finitelyEnumerable : FinitelyEnumerable Apex
+    inj : Injection X Apex
```

Comment:
@JacquesCarette I can *just about* take `Apex` as the name of the 'covering object' `A` in an epic situation `A → X` (L52-L54 below), but it seems less defensible here in the dual situation `X → A`. Is it 'standard' terminology? Or is there something better from eg the categorical literature?

Reply:
Yes, this is probably too influenced by other code / ideas, and no longer really fits here. Suggestions?

Reply:
I hunted around, and unfortunately the literature uses 'generic' names such as `X`, `Y`/`A`, `B` when discussing things as morphisms. For embedding/extension, the domain of an injection is sometimes called `I`, and of a projection as `P`, but these don't seem quite satisfactory, either?

In each case, though, the relevant object is (suitably) enumerable, so could it be called `Enum`, or even `Enumeration`?

Reply:
`Enumeration` > `Enum` > `E`.